akka-projection-jdbc/src/main/scala/akka/projection/jdbc/internal/Dialect.scala

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.jdbc.internal
+
+trait Dialect {
+  val createTableStatement: String
+  val alterTableStatement: String
+
+  val insertOrUpdateStatement: String
+  val readOffsetQuery: String
+  val clearOffsetStatement: String
+
+}
+
+/* Defines the basic statements. Dialects may use it or define their own.
+ */
+object DialectDefaults {
+  def readOffsetQuery(table: String) =
+    s"SELECT * FROM $table WHERE PROJECTION_NAME = ?"
+
+  def clearOffsetStatement(table: String) =
+    s"DELETE FROM $table WHERE projection_name = ? AND projection_key = ?"
+
+  def insertOrUpdateStatement(table: String) =
+    s"""MERGE INTO "$table" ("PROJECTION_NAME","PROJECTION_KEY","OFFSET","MANIFEST","MERGEABLE","LAST_UPDATED")  VALUES (?,?,?,?,?,?)"""
+
+  // NOTE: keep this aligned with insertOrUpdateStatement columns
+  object InsertIndices {
+    val PROJECTION_NAME = 1
+    val PROJECTION_KEY = 2
+    val OFFSET = 3
+    val MANIFEST = 4
+    val MERGEABLE = 5
+    val LAST_UPDATED = 6
+  }
+
```

Comment:
Is there any way to make this less errorprone?  Maybe we could write setter methods per statement, per dialect.

Reply:
yes, I have been thinking about it. I will revisit when we add more dialects. I expect to move a few things around when adding Postgres. The statement for `insertOrUpdate` in Postgres will probably have more indices.